### ISSUE TITLE
BN Tuning

### ISSUE THREAD

**Issue by cfiliu** [zhoudw-zdw/RevisitingCIL#3]
Hi!

First of all, wonderful work, it's so interesting and useful. I'm working with CNN and I'm interested in Batch Normalization Tuning for CIL, but I can't find its implementation on code, something like "adam_bn_tuning" in the "models" folder.

I don't know if it is in another part of the code, but I'm not able to find it. It would be fantastic if you could help me with this issue.

Thanks in advance!

**Comment by zhoudw-zdw** [comment]
Thanks for your interest.

Implementing BN tuning is very easy. You just need to clear BN statistics and forward the training instances (in the training mode). Here is a simple demo for the first step:

```
def clear_running_mean(self):
        # record the index of running mean and variance 
        model_dict=self._network.state_dict()
        running_dict = {}
        for e in model_dict:
            if 'running' in e:
                key_name = '.'.join(e.split('.')[1:-1])
                if key_name in running_dict:
                    continue
                else:
                    # print('running name',key_name)
                    running_dict[key_name] = {}
                # find the position of BN modules
                component = self._network.convnet
                for att in key_name.split('.'):
                    # print(att)
                    if att.isdigit():
                        component = component[int(att)]
                    else:
                        component = getattr(component, att)
                
                running_dict[key_name]['mean'] = component.running_mean
                running_dict[key_name]['var'] = component.running_var
                running_dict[key_name]['nbt'] = component.num_batches_tracked

                component.running_mean = component.running_mean*0
                component.running_var = component.running_var*0
                component.num_batches_tracked = component.num_batches_tracked*0

        print(component.running_mean,component.running_var,component.num_batches_tracked)
```

I will include a full version in the next update, but currently, I am a little busy and cannot promise a timeline.
